worlds/tunic/__init__.py

```diff
@@ -321,8 +347,53 @@ def remove_filler(amount: int) -> None:
             if tunic_item.name in slot_data_item_names:
                 self.slot_data_items.append(tunic_item)
 
+        # pull out the filler so that we can place it manually during pre_fill
+        self.local_filler = []
+        if self.options.grass_randomizer and self.options.local_fill > 0 and self.multiworld.players > 1:
+            # skip items marked local or non-local, let fill deal with them in its own way
+            # discard grass from non_local if it's meant to be limited
+            if self.settings.limit_grass_rando:
+                self.options.non_local_items.value.discard("Grass")
+            all_filler: List[TunicItem] = []
+            non_filler: List[TunicItem] = []
+            for tunic_item in tunic_items:
+                if (tunic_item.classification in [ItemClassification.filler, ItemClassification.trap]
+                        and tunic_item.name not in self.options.local_items
+                        and tunic_item.name not in self.options.non_local_items):
+                    all_filler.append(tunic_item)
+                else:
+                    non_filler.append(tunic_item)
+            amount_to_local_fill = int(self.options.local_fill.value * len(all_filler) / 100)
+            self.local_filler = all_filler[:amount_to_local_fill]
+            del all_filler[:amount_to_local_fill]
+            tunic_items = all_filler + non_filler
+
         self.multiworld.itempool += tunic_items
 
+    @classmethod
+    def stage_pre_fill(cls, multiworld: MultiWorld) -> None:
+        tunic_grass_worlds: List[TunicWorld] = [world for world in multiworld.get_game_worlds("TUNIC")
+                                                if world.options.grass_randomizer]
+        tunic_players_with_grass: List[int] = [world.player for world in tunic_grass_worlds]
+        # we need to reserve a couple locations so that we don't fill up every sphere 1 location
+        reserved_locations: Set[str] = set(multiworld.random.sample(sphere_one, 2))
+        unfilled_locations = [loc for loc in multiworld.get_unfilled_locations_for_players(
+            location_names=[], players=tunic_players_with_grass) if loc.progress_type != LocationProgressType.PRIORITY
+                              and loc.name not in reserved_locations]
+        grass_filler_items: List[TunicItem] = []
+        for world in tunic_grass_worlds:
+            grass_filler_items.extend(world.local_filler)
```

Comment:
You could reorder this part and add an early return just so that you aren't calling `get_unfilled_locations_for_players` when it isn't needed.
```suggestion
        grass_filler_items: List[TunicItem] = []
        for world in tunic_grass_worlds:
            grass_filler_items.extend(world.local_filler)

        if not grass_filler_items:
            return

        unfilled_locations = [loc for loc in multiworld.get_unfilled_locations_for_players(
            location_names=[], players=tunic_players_with_grass) if loc.progress_type != LocationProgressType.PRIORITY
                              and loc.name not in reserved_locations]
```